Adds bloom filter aggregator to 'druid-bloom-filters' extension

clintropolis opened the pull request:
This PR, building on top of the work introduced in #6222, extends `druid-bloom-filters` with a query time aggregator, allowing bloom filters to be computed from query results which can then be used as input to `bloom` filters in subsequent queries.

example query:
```json
{
  "queryType": "timeseries",
  "dataSource": "wikiticker",
  "intervals": [ "2015-09-12T00:00:00.000/2015-09-13T00:00:00.000" ],
  "granularity": "day",
  "aggregations": [
    {
      "type": "bloom",
      "name": "userBloom",
      "maxNumEntries": 100000,
      "field": {
        "type":"default",
        "dimension":"user",
        "outputType": "STRING"
      }
    }
  ]
}
```
example results:
```json
[{"timestamp":"2015-09-12T00:00:00.000Z","result":{"userBloom":"BAAAJhAAAA..."}}]
```

dclim commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeBufferAggregator.java:
```diff
@@ -0,0 +1,128 @@
+/*
```
There's quite a lot of duplicate or similar code between this and BloomFilterBufferAggregator. Any opportunity to consolidate?

clintropolis replied:
Consolidated common code into `BaseBloomFilterAggregator` and `BaseBloomFilterBufferAggregator`

leventov commented on processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -99,6 +99,9 @@
   public static final byte STRING_FIRST_CACHE_TYPE_ID = 0x2B;
   public static final byte STRING_LAST_CACHE_TYPE_ID = 0x2C;
 
+  public static final byte BLOOM_FILTER_CACHE_TYPE_ID = 0x30;
```
0x30 and 0x31 are already in use. Also position those variables in the end of the list.

clintropolis replied:
Oops, originally did this work in an older branch where these didn't exist yet I think, will fix.

leventov replied:
Could you make this kind of mistake impossible in the future by moving all codes in a enum with `byte id;` field, and adding a static initializer that checks that all enum constants have different codes?

Also `CacheKeyBuilder`'s constructor to accept the enum constant instead of `byte` param to prohibit bypassing this enum

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/types/StringBloomFilterAggregatorColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,49 @@
+  public void add(DimensionSelector selector, BloomKFilter bloomFilter)
+  {
+    if (selector.getRow().size() > 1) {
+      String[] strings = (String[]) selector.getObject();
```
Could avoid creating an array by iterating the row itself.

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/types/BloomFilterAggregatorColumnSelectorStrategyFactory.java:
```diff
@@ -0,0 +1,50 @@
+{
+  @Override
+  public BloomFilterAggregatorColumnSelectorStrategy makeColumnSelectorStrategy(
+      ColumnCapabilities capabilities, ColumnValueSelector selector
```
Formatting

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/types/DoubleBloomFilterAggregatorColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,36 @@
+import org.apache.hive.common.util.BloomKFilter;
+
+public class DoubleBloomFilterAggregatorColumnSelectorStrategy
+    implements BloomFilterAggregatorColumnSelectorStrategy<DoubleColumnSelector>
```
`DoubleColumnSelector` must appear only in `implements` clauses and nowhere else. See it's Javadoc. Same for float and long.

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeBufferAggregator.java:
```diff
@@ -0,0 +1,128 @@
+  @Override
+  public void close()
+  {
+
```
Please add comment "nothing to close"

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeBufferAggregator.java:
```diff
@@ -0,0 +1,128 @@
+    final int oldLimit = buf.limit();
+    try {
+      buf.position(position);
+      BloomKFilter collector = BloomKFilter.deserialize(new ByteBufferInputStream(buf));
```
It there a way to update Bloom Filter in place without streaming deserialization and than later serialization back?

clintropolis replied:
There is a static method, [`BloomKFilter.mergeBloomFilterBytes`](https://github.com/apache/hive/blob/master/storage-api/src/java/org/apache/hive/common/util/BloomKFilter.java#L360) which can do a merge of 2 bloom filters directly, so this could be done for the merge operation at least, will look into it.

clintropolis replied:
Added `ByteBuffer` methods to our copied `BloomKFilter` to add items to and merge serialized bloom filters in place, so buffer aggs should be much more chill to the heap

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeBufferAggregator.java:
```diff
@@ -0,0 +1,128 @@
+      BloomKFilter collector = BloomKFilter.deserialize(new ByteBufferInputStream(mutationBuffer));
+      return collector;
+    }
+    catch (Exception ex) {
```
Couldn't narrow the exception type?

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeBufferAggregator.java:
```diff
@@ -0,0 +1,128 @@
+        BloomKFilter.serialize(out, collector);
+      }
+    }
+    catch (Exception ex) {
```
Couldn't narrow the exception type?

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeBufferAggregator.java:
```diff
@@ -0,0 +1,128 @@
+    try {
+      BloomKFilter.serialize(outputStream, filter);
+    }
+    catch (Exception ex) {
```
Couldn't narrow the exception type?

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeAggregatorFactory.java:
```diff
@@ -0,0 +1,85 @@
+  private String fieldName;
+
+  BloomFilterMergeAggregatorFactory(
+      String name,
```
Unnecessary breakdown

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeAggregatorFactory.java:
```diff
@@ -0,0 +1,85 @@
+  {
+    final ColumnValueSelector<BloomKFilter> selector = metricFactory.makeColumnValueSelector(fieldName);
+    if (selector instanceof NilColumnValueSelector) {
+      return NoopAggregator.instance();
```
I wonder if this valid? Null and empty Bloom filter is not the same thing

clintropolis replied:
I think that this shouldn't happen, rather nulls values will still result in a `BloomKFilter`, here so changed this to throw an exception.

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeAggregator.java:
```diff
@@ -0,0 +1,82 @@
+  {
+    return collector;
+  }
+
```
extra line

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeAggregator.java:
```diff
@@ -0,0 +1,82 @@
+  @Override
+  public void close()
+  {
+
```
Please add comment "nothing to close"

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeAggregator.java:
```diff
@@ -0,0 +1,82 @@
+  private BloomKFilter collector;
+
+  public BloomFilterMergeAggregator(
+      ColumnValueSelector<BloomKFilter> selector,
```
Unnecessary breakdown

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterBufferAggregator.java:
```diff
@@ -0,0 +1,127 @@
+import java.nio.ByteBuffer;
+
+
+public class BloomFilterBufferAggregator implements BufferAggregator
```
Same comments as to other similar classes in this PR

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
```
`@Nullable` params

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,233 @@
+  public BloomFilterAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("field") final DimensionSpec field,
+      @JsonProperty("maxNumEntries") Integer maxNumEntries
```
`@Nullable`

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,233 @@
+  {
+    this.name = name;
+    this.field = field;
+    this.maxNumEntries = maxNumEntries != null ? maxNumEntries : 1500;
```
Please make a constant

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregator.java:
```diff
@@ -0,0 +1,79 @@
+  @Override
+  public void close()
+  {
+
```
Please add a comment "nothing to close"

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/types/BloomFilterAggregatorColumnSelectorStrategyFactory.java:
```diff
@@ -0,0 +1,50 @@
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ValueType;
+
+public class BloomFilterAggregatorColumnSelectorStrategyFactory
```
Is it sure that "Aggregator" should be a part of the name of this class?

clintropolis replied:
I think it's not necessary, though I was just following [cardinality aggregator](https://github.com/apache/incubator-druid/blob/master/processing/src/main/java/org/apache/druid/query/aggregation/cardinality/types/CardinalityAggregatorColumnSelectorStrategyFactory.java).

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,233 @@
+  @Override
+  public int getMaxIntermediateSize()
+  {
+    BloomKFilter throwaway = new BloomKFilter(maxNumEntries);
```
Cache the value in a static, don't create a throwaway each time.

clintropolis replied:
What sort of cache is most appropriate? A static int2object map? caffeine?

leventov replied:
"cache" = set to a static final field, I meant here

clintropolis replied:
I don't think a static works since the result depends on `maxNumEntries` which is a query time parameter. If this method gets called multiple times i can set an instance field to this value at constructor time, or I can see if the math that computes the size of the long array inside is accessible to call that directly.

leventov replied:
Direct computation, if not supported by the library directly (maybe we could contribute that?) could be dangerous if the algorithm in the library changes in some version.

I think most of the time only one `maxNumEntries` value will be seen per JVM, small Int2Object map (stopped being populated after say 10 entires) should work

leventov replied:
*Int2IntMap, you could cache the final values.

clintropolis replied:
Added a method to our copied `BloomKFilter` to compute the size required given a number of entries, avoiding this throwaway

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,233 @@
+  public int getMaxIntermediateSize()
+  {
+    BloomKFilter throwaway = new BloomKFilter(maxNumEntries);
+    return (throwaway.getBitSet().length * Long.BYTES) + 5;
```
What is +5?

clintropolis replied:
It's a header written during serialization I'll add a comment/link to https://github.com/apache/hive/blob/master/storage-api/src/java/org/apache/hive/common/util/BloomKFilter.java#L302 or see if I can find a better way to get this information

clintropolis replied:
Now uses new method on copied `BloomKFilter` to compute the size

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,233 @@
+  public Comparator getComparator()
+  {
+    // idk how to compare?
+    return (Comparator<BloomKFilter>) (o1, o2) -> 0;
```
Could compare the number of set bits.

clintropolis replied:
:+1:

clintropolis replied:
Added method to copied `BloomKFilter` to count the number of set bits, allowing ordering results by density of bloom filter

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterBufferAggregator.java:
```diff
@@ -0,0 +1,127 @@
+    final ByteBuffer mutationBuffer = buf.duplicate();
+    mutationBuffer.position(position);
+    BloomKFilter filter = new BloomKFilter(maxNumEntries);
+    ByteBufferBackedOutputStream wat = new ByteBufferBackedOutputStream(mutationBuffer);
```
What does this variable name mean?

clintropolis replied:
Heh, oops, missed some clean up before opening the PR

clintropolis commented on docs/content/development/extensions-core/bloom-filter.md:
````diff
@@ -4,21 +4,29 @@ layout: doc_page
+### JSON Specification of Bloom Filter
 ```json
 {
   "type" : "bloom",
````
I believe that this is really only ugly on github and it looks ok translated to the website docs

clintropolis commented on docs/content/development/extensions-core/bloom-filter.md:
```diff
@@ -42,4 +50,53 @@ Internally, this implementation of bloom filter uses Murmur3 fast non-cryptograp
+|`type`                   |Aggregator Type. Should always be `bloom`|yes|
+|`name`                   |Output field name |yes|
+|`field`                  |[DimensionSpec](./../dimensionspecs.html) to add to `org.apache.hive.common.util.BloomKFilter` | yes |
+|`maxNumEntries`          |Maximum number of distinct values supported by `org.apache.hive.common.util.BloomKFilter`, default `1500`| no |
```
Hmm, digging into it, in `BloomKFilter` the [false positive rate is not controllable](https://github.com/apache/hive/blob/master/storage-api/src/java/org/apache/hive/common/util/BloomKFilter.java#L61) in the manner of `BloomFilter`, and is fixed to the default of 5%. However I guess that can be indirectly controlled by increasing the `maxNumEntries`, though that's kind of lame. Having a higher cardinality than the value of `maxNumEntries` will cause the false positive probability to reach 1, constructing a useless bloom filter, so that should definitely be added to the docs.

clintropolis replied:
Updated docs to include fixed 5% false positive rate, though no formula for how changing `maxNumEntries` affects that yet.

clintropolis commented on docs/content/development/extensions-core/bloom-filter.md:
```diff
@@ -42,4 +50,53 @@ Internally, this implementation of bloom filter uses Murmur3 fast non-cryptograp
+Note: `org.apache.hive.common.util.BloomKFilter` provides a serialize method which can be used to serialize bloom filters to outputStream.
+
+## Bloom Filter Query Aggregator
+Input for a `bloomKFilter` can also be created from a druid query with the `bloom` aggregator.
```
`bloom` is the correct value to be consistent with the filter type name, updated docs to reflect that.

clintropolis commented:
This PR is dependent on #6546 being resolved.

clintropolis commented:
@dclim, @leventov (and anyone else interested) I think this PR is ready for review again. I think I've addressed all existing comments via code changes, apologies for the rebases - I jumped the gun a bit opening this before the rest of the bloom filter extension had stabilized. 

I've also added a bunch of methods to our copy of the Hive BloomKFilter to enable in situ manipulation of BloomKFilters that are serialized into a ByteBuffer, to allow much more memory efficient buffer aggs, which I will attempt to get pushed to the upstream implementation soon to avoid diverging too much. Additionally the documentation has been improved a bit for the extension, and have simplified the code where possible in response to comments.

clintropolis commented:
Oops, have some legitimate test failures after a few changes, fixing.

jihoonson commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,256 @@
+  }
+
+  @Override
+  public Comparator getComparator()
```
Would you please document the sort order for bloom filters?

jihoonson commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/filter/BloomKFilter.java:
```diff
@@ -210,6 +218,214 @@ public static void mergeBloomFilterBytes(
+     * Serialized BloomKFilter format:
+     * 1 byte for the number of hash functions.
+     * 1 big endian int(That is how OutputStream works) for the number of longs in the bitset
+     * big endina longs in the BloomKFilter bitset
```
endina -> endian

jon-wei commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,256 @@
+  @Override
+  public Comparator getComparator()
+  {
+    return (Comparator<Object>) (o1, o2) -> {
```
this comparator looks fairly expensive, let's mention that in the docs

clintropolis replied:
Added docs, also improved the set bit counting so it shouldn't be quite as expensive as it was previously.

clintropolis replied:
Wrapped with `Comparators.nullsFirst`

jon-wei commented on docs/content/development/extensions-core/bloom-filter.md:
```diff
@@ -24,22 +24,43 @@ title: "Bloom Filter"
+to use with Druid for cases where an explicit filter is impossible, e.g. filtering a query against a set of millions of
+ values.
+ 
+Following are some characterstics of BloomFilters:
```
characterstics -> characteristics

jon-wei commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/filter/BloomKFilter.java:
```diff
@@ -210,6 +218,214 @@ public static void mergeBloomFilterBytes(
+  }
+
+  /**
+   * Caculate size in bytes of a BloomKFilter for a given number of entries
```
Caculate -> Calculate

clintropolis replied:
Fixed (and a few other occurrences)

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/guice/BloomFilterExtensionModule.java:
```diff
@@ -40,6 +40,8 @@
   public void configure(Binder binder)
   {
-    SqlBindings.addOperatorConversion(binder, BloomFilterOperatorConversion.class);
+    if (binder != null) {
```
This is only necessary because `BloomFilterGroupByQueryTest`, will see if i can rework to not be required, or add a comment for why it's needed.

clintropolis replied:
Changed to not be needed and removed null check

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterBufferAggregator.java:
```diff
@@ -0,0 +1,56 @@
+  {
+    final int oldPosition = buf.position();
+    buf.position(position);
+    selectorPlus.getColumnSelectorStrategy().bufferAdd(selectorPlus.getSelector(), buf);
```
Added striped lock for get/aggregate as per linked threads

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BaseBloomFilterBufferAggregator.java:
```diff
@@ -0,0 +1,78 @@
+  {
+    ByteBuffer mutationBuffer = buf.duplicate();
+    mutationBuffer.position(position);
+    int sizeBytes = 5 + (buf.getInt(position + 1) << 3);
```
added comment explaining layout

clintropolis commented:
@leventov any additional comments before I fix up conflicts?

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BaseBloomFilterBufferAggregator.java:
```diff
@@ -0,0 +1,93 @@
+      // | k (byte) | numLongs (int) | bitset (long[numLongs]) |
+      int sizeBytes = 1 + Integer.BYTES + (buf.getInt(position + 1) * Long.BYTES);
+      mutationBuffer.limit(position + sizeBytes);
+      return mutationBuffer.slice();
```
I'm afraid that this comment: https://github.com/apache/incubator-druid/pull/5148#discussion_r171968971 does actually apply to this aggregator, i. e. synchronization is pointless without changes in the aggregation system as a whole.

clintropolis replied:
Removed for now in favor of resolving this issue in system level

gianm commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/filter/BloomKFilter.java:
```diff
@@ -38,7 +40,13 @@
+ * ByteBuffer, e.g. all add and merge methods. Test methods were not added because we don't need them.. but would
+ * probably be chill to do so it is symmetrical.
+ *
+ * Todo: remove this and begin using hive-storage-api version again once
```
The original motivation for this extension was for Hive integration: https://github.com/apache/incubator-druid/pull/6222. So from that perspective it should remain compatible with Hive's format. But I don't see any issue with offering support for two formats: one compatible with Hive and one optimized for what we want to do.

clintropolis replied:
A future improvement I would like to do is adding a `type` field to the filter and aggregator of this extension to allow exploring other backing bloom filter or hash function implementations and/or different storage formats while still maintaining a hive compatible option.

clintropolis replied:
As I understand the Hive integration is that it externally constructs a `BloomKFilter` with a bunch of values and then uses it with the `BloomDimFilter` the original PR added to filter Druid queries. 

This agg uses the same bloom filter implementation so there is a native Druid way to generate input for `BloomDimFilter` from existing Druid data.

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/filter/BloomKFilter.java:
```diff
@@ -210,6 +220,261 @@ public static void mergeBloomFilterBytes(
+      return new BloomKFilter(data, numHashFunc);
+    }
+    catch (RuntimeException e) {
+      IOException io = new IOException("Unable to deserialize BloomKFilter");
```
Fixed, I had just copied the instance methods as the basis of the `ByteBuffer` methods and that's how it did this part.

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterBufferAggregator.java:
```diff
@@ -0,0 +1,56 @@
+  @Override
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    inspector.visit("selectorPlus", selectorPlus);
```
:+1:

`CardinalityBufferAggregator` has an array of `SelectorPlus` so modified it to loop and visit each selector of the selector plus, lmk if something else would be better.

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,253 @@
+  @Override
+  public Comparator getComparator()
+  {
+    return Comparator.nullsFirst((o1, o2) -> {
```
Fixed

clintropolis commented:
Any more comments on this PR @leventov ?

clintropolis commented:
@leventov would you mind if I merge this soon so I can unblock another branch I'm sitting on that adds sql support on top of this? If you find any additional issues I am happy to address them in a future patch.

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterBufferAggregator.java:
```diff
@@ -0,0 +1,56 @@
+  @Override
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    inspector.visit("selector", selectorPlus.getSelector());
```
Oops, forgot about the case statement, will add :+1:

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeAggregatorFactory.java:
```diff
@@ -0,0 +1,81 @@
+  {
+    final ColumnValueSelector<BloomKFilter> selector = metricFactory.makeColumnValueSelector(fieldName);
+    if (selector instanceof NilColumnValueSelector) {
+      throw new ISE("WTF?! Unexpected NilColumnValueSelector");
```
[To expand on this comment](https://github.com/apache/incubator-druid/pull/6397#discussion_r247308520), when examining how this would be called my conclusion was that this would be an unexpected condition in the merge aggregator, because the output of a null column from the aggregator would still be a bloom filter with the null or default value which is what this would see. Did I misinterpret or miss a situation where this could actually happen?

leventov replied:
I just noticed that this is the merge aggregator. Then it probably makes sense, but there should be a comment and the error message should be more descriptive.

But then it appears that `BloomFilterAggregatorFactory` doesn't special-case `NilColumnValueSelector`, however it probably should for performance.

clintropolis replied:
After refactor this now special cases `NilColumnValueSelector` and I think behaves in a slightly more correct manner of producing a totally empty bloom filter instead of a bloom filter with a null value

leventov replied:
Please add a comment

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/types/BloomFilterAggregatorColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.nio.ByteBuffer;
+
+public interface BloomFilterAggregatorColumnSelectorStrategy<TValueSelector> extends ColumnSelectorStrategy
```
I used the same abstraction as `CardinalityAggregator`, should it be refactored in this manner as well? (not in this PR, but later at least)

To clarify you're suggesting something like having factorize produce a `StringBloomFilterAggregator`, `LongBloomFilterAggregator`, etc with the same for buffer aggs? I'll have a look at reworking it.

leventov replied:
Yes, exactly. I think this Strategy and StrategyFactory is too much and the logic is lost between endless classes.

Yes, I think cardinality should be refactored too. I've created #6909.

clintropolis replied:
Thanks :+1:

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/types/StringBloomFilterAggregatorColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,73 @@
+  {
+    if (selector.getRow().size() > 1) {
+      selector.getRow().forEach(v -> {
+        String value = selector.lookupName(v);
```
Javadocs aren't super clear and it isn't obvious to me yet looking at code, does `nameLookupPossibleInAdvance` indicate that indexes uniquely map to the same value across all segments? That would need to be true for this to work, and it would also require modifications to the `BloomDimFilter` implementation to be able to work with this as well.

I think the main use case of this aggregator right now is to be able to produce bloom filters from data in druid, which can be used as input to additional queries to filter other data in druid using `BloomDimFilter` which this extension originally introduced, so the goal is to have them operate in the same manner.

It definitely is expensive to do the hashing, it feels even apparent on the `BloomDimFilter` side of things, but I think this optimization is maybe out of scope of this PR

leventov replied:
> Javadocs aren't super clear and it isn't obvious to me yet looking at code, does nameLookupPossibleInAdvance indicate that indexes uniquely map to the same value across all segments? That would need to be true for this to work, and it would also require modifications to the BloomDimFilter implementation to be able to work with this as well.

Yes, map uniquely. Feel free to clarify that Javadoc right in this PR.

> I think the main use case of this aggregator right now is to be able to produce bloom filters from data in druid, which can be used as input to additional queries to filter other data in druid using BloomDimFilter which this extension originally introduced, so the goal is to have them operate in the same manner.

> It definitely is expensive to do the hashing, it feels even apparent on the BloomDimFilter side of things, but I think this optimization is maybe out of scope of this PR

OK, then please leave a comment somewhere with an explanation why the optimization is not applied.

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -89,27 +87,69 @@ public BloomFilterAggregatorFactory(
 
-    return new BloomFilterAggregator(selectorPlus, maxNumEntries);
+    if (selector instanceof NilColumnValueSelector) {
+      return new NilBloomFilterAggregator((NilColumnValueSelector) selector, filter);
```
Is it important that `BloomKFilter` has specifically `maxNumEntries`? If not, `new BloomKFilter(0)` (or 1) could be cached in a constant.

leventov replied:
In any case, `(NilColumnValueSelector) selector` shouldn't be passed down, `NilColumnValueSelector` could call `super(NilColumnValueSelector.instance())` in its constructor.

clintropolis replied:
`BloomKFilter` must be the same size to merge, so it's not possible to make a constant. :+1: on the signature change though, will do that.

leventov replied:
Ok, please add a comment noting this.

clintropolis commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,296 @@
+import java.util.List;
+import java.util.Objects;
+
+public class BloomFilterAggregatorFactory extends AggregatorFactory
```
As far as I could tell while implementing this, `makeAggregateCombiner` is only called to merge segments at ingestion time; the issue that those pulls reference, #6877, maybe jives with that since it is an exception during index merging.

leventov replied:
The long term idea is to replace the remaining uses of `combine()` with `AggregateCombiner` and remove `combine()` method, to reduce repetition. When this time comes, it will be harder to implement that for every aggregator. So IMO it's better to implement `makeAggregateCombiner()` in all aggregators in core eagerly.

clintropolis replied:
Added `BloomFilterAggregateCombiner`

leventov commented on extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterAggregatorFactory.java:
```diff
@@ -0,0 +1,296 @@
+    if (capabilities == null) {
+      BaseNullableColumnValueSelector selector = columnFactory.makeColumnValueSelector(field.getDimension());
+      if (selector instanceof NilColumnValueSelector) {
+        return new NilBloomFilterAggregator((NilColumnValueSelector) selector, filter);
```
Could you please rename "NoOp" aggregators (I see it in ArrayOfDoublesSketch), and whatever other nonstandard names other complex aggregators use (I only see that DistinctCount already uses "Empty") and align everything to the same convention?

clintropolis replied:
Created https://github.com/apache/incubator-druid/issues/6934 and assigned to myself to do as a follow-up PR

jon-wei commented:
Still LGTM after latest changes

clintropolis commented:
Thanks for review everyone :metal:
